lib/posix-process/signal/deliver.c

```diff
@@ -0,0 +1,352 @@
+{
+	struct posix_process *pproc;
+	uk_sigset_t saved_mask;
+	int signum;
+
+	UK_ASSERT(sig && sig->siginfo.si_signo);
+	UK_ASSERT(pthread);
+
+	pproc = tid2pprocess(pthread->tid);
+	UK_ASSERT(pproc);
+
+	signum = sig->siginfo.si_signo;
+
+	/* Execute the default action if a signal handler is not defined */
+	if (KERN_SIGACTION(pproc, signum)->ks_handler == SIG_DFL) {
+		/* Standard signals: Invoke signal-specific default action */
+		if (signum < SIGRTMIN) {
+			if (UK_BIT(signum) & SIGACT_CORE_MASK)
+				uk_sigact_core(signum);
+			else if (UK_BIT(signum) & SIGACT_TERM_MASK)
+				uk_sigact_term(signum);
+			else if (UK_BIT(signum) & SIGACT_STOP_MASK)
+				uk_sigact_stop(signum);
+			else if (UK_BIT(signum) & SIGACT_CONT_MASK)
+				uk_sigact_cont(signum);
+			else if (UK_BIT(signum) & SIGACT_IGN_MASK)
+				uk_sigact_ign(signum);
+			else
+				UK_BUG();
+		} else {
+			/* Real-time signals: SIG_TERM by default */
+			uk_sigact_term(signum);
+		}
+		return 0;
+	}
+
+	/* Save original mask and apply mask from sigaction */
+	saved_mask = pthread->signal->mask;
+	uk_sigorset(&pthread->signal->mask,
+		    &KERN_SIGACTION(pproc, signum)->ks_mask);
+
+	/* Also add this signal to masked signals */
+	if (!(KERN_SIGACTION(pproc, signum)->ks_flags & SA_NODEFER))
+		SET_MASKED(pthread, signum);
+
+	/* Execute handler */
+	handle_self(sig, KERN_SIGACTION(pproc, signum), execenv);
+
+	/* Restore original mask */
+	pthread->signal->mask = saved_mask;
+
+	/* If SA_RESETHAND flag is set, restore the default handler */
+	if (KERN_SIGACTION(pproc, signum)->ks_flags & SA_RESETHAND)
+		pprocess_signal_sigaction_clear(KERN_SIGACTION(pproc, signum));
+
+	return 0;
+}
+
+/* Deliver pending signals of a given process. We deliver each
+ * pending signal to the first thread that doesn't mask that
+ * signal.
+ */
+static int deliver_pending_proc(struct posix_process *proc,
+				struct ukarch_execenv *execenv)
+{
+	struct posix_thread *thread, *threadn;
+	struct posix_thread *this_thread;
+	struct uk_signal *sig;
+	int handled_cnt = 0;
+	bool handled;
+
+	UK_ASSERT(proc);
+	UK_ASSERT(execenv);
+
+	this_thread = uk_pthread_current();
+	UK_ASSERT(this_thread);
+
+	for (int i = 0; i < SIG_ARRAY_COUNT; i++) {
+		handled = false;
+
+		/* Skip if the process ignores this signal altogether */
+		if (IS_IGNORED(proc, i))
+			continue;
+
+		/* POSIX specifies that if a signal targets the
+		 * current process / thread, then at least one
+		 * signal for this process /thread must be
+		 * delivered before the syscall returns, as long as:
+		 *
+		 * 1. No other thread has that signal unblocked
+		 * 2. No other thread is in sigwait() for that signal (TODO)
+		 */
+		uk_pprocess_foreach_pthread(proc, thread, threadn) {
+			if (thread->tid == this_thread->tid)
+				continue;
+
+			if (IS_MASKED(thread, i))
+				continue;
+
+			while ((sig = pprocess_signal_dequeue(proc, NULL, i))) {
+				do_deliver(thread, sig, execenv);
+				uk_signal_free(proc->_a, sig);
+				handled = true;
```

Comment:
Do we still need `handled` if we can check `handled_cnt > 0` in the if below?

Reply:
No there is some redundancy here, but if we remove `handled_cnt` in the end (as discussed in your previous comment), `handled` will be our only choice. Let's address all these once we merge shutdown.